### PR TITLE
Add Support to compute Sample Perplexity

### PR DESCRIPTION
**Pull request by vbaddi**
This PR adds a script for calculating the perplexity of ONNX, QPC, or Torch models using the WikiText-2 dataset.

`Key Features:`

1. Supports ONNX, QPC, and Torch models.
2. Configurable parameters for model path, context length, prompt length, batch size, and more.
3. Logs perplexity, loss, and evaluation time.
4. Example usage provided in the README.md.

### CODE REVIEW

File: scripts/README.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+- Install QEfficient and update the datasets package to 2.20
+
+## Usage
+
+To run the script, use the following command:
+
+```bash
+python perplexity_calculator_cloud.py --model_type <model_type> --model_name <model_name> [--model_path <model_path>] [--dataset_name <dataset_name>] [--ctx_len <ctx_len>] [--prompt_len <prompt_len>] [--batch_size <batch_size>] [--stride <stride>] [--num_samples <num_samples>] [--qpc_device_id <qpc_device_id>] [--log_file <log_file>]
+
+python perplexity_calculator_cloud.py --model_type torch --model_name meta-llama/Meta-Llama-3-8B-Instruct --num_samples 1
```
Comment (quic-akuruvil) [inline scripts/README.md:32]: `perplexity_calculator_cloud.py` filename is missing.

Comment (vbaddi) [reply scripts/README.md:32]: Done

File: scripts/calculate_perplexity.py [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+                if len(logits.shape) == 2:
+                    logits = np.expand_dims(logits, axis=1)
+                if logits.dtype == np.float16:
+                    logits = logits.astype(np.float32)
+
+                inputs["input_ids"] = inp_ids[:, cache_index]
+                targets_label.append(torch.from_numpy(inputs["input_ids"]))
+                inputs["position_ids"] += 1
+
+                if not is_qaic:
```
Comment (quic-akuruvil) [inline scripts/calculate_perplexity.py:298]: For `onnx` type, the onnx model generated from Qeff has large constants like `2147483647`. This needs to be fixed too, before running on onnxrt.

Comment (quic-akuruvil) [reply scripts/calculate_perplexity.py:298]: Due to invalid indices, are we fixing this elsewhere?

File: scripts/calculate_perplexity.py [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+                cache_index += 1
+                inputs["input_ids"] = inp_ids[:, cache_index]
+                targets_label.append(torch.from_numpy(inputs["input_ids"]))
+                inputs["position_ids"] = input_len
+            else:
+                outputs = {}
+                input_names, input_shapes, output_names = inference_session._input_output_specs()
+                for i, iname in enumerate(input_names):
+                    if "past" in iname:
+                        inputs[iname] = np.zeros(
```
Comment (quic-akuruvil) [inline scripts/calculate_perplexity.py:275]: For models with 3D KV layout (like starcoder1 architecture) one more case need to be added.

Comment (vbaddi) [reply scripts/calculate_perplexity.py:275]: @quic-akuruvil , can you push this change on the same PR pls.
